storage/innobase/include/mem0mem.ic

```diff
@@ -459,6 +461,8 @@ mem_heap_free_top(
 		Originally, this was UNIV_MEM_FREE(), to catch writes
 		to freed memory. */
 		UNIV_MEM_ALLOC((byte*) block + mem_block_get_free(block), n);
+		ASAN_POISON_MEMORY_REGION(
+			(byte*)block + mem_block_get_free(block), n);
```

Comment:
I would suggest to change the `UNIV_MEM_ALLOC` to `UNIV_MEM_FREE` and make `ASAN_POISON_MEMORY_REGION` a part of that.

Similarly, consider making `ASAN_UNPOISION_MEMORY_REGION` a part of `UNIV_MEM_ALLOC`.
If we only want to declare memory uninitialized (which would be useful with MSAN) then `UNIV_MEM_UNDEFINED` can be used for that. Currently under Valgrind it seems to be equivalent to `UNIV_MEM_ALLOC`, but there is no reason for it to be like that.

Reply:
I think that we should do the poisoning and unpoisoning in `Pool::mem_free()` and `Pool::get()` as well, to catch use of `trx_t` objects after `trx_free()`. Adding Valgrind `UNIV_MEM_UNDEFINED` probably would not work, because I think that the intention is to not initialize certain fields that are supposed to be at ‘default values’ when the object is stored in the pool.

Reply:
> If we only want to declare memory uninitialized (which would be useful with MSAN) then UNIV_MEM_UNDEFINED can be used for that.

MSAN is a total different story. I've even never tried it because it requires at least building instrumented standard library.